File: ProjectEulerPython/problems/problem_034.py
from problems.problem import Problem

def digit_factorial(d: int) -> int:
  if d == 0:
    return 1
  if d == 1:
    return 1
  if d == 2:
    return 2
  if d == 3:
    return 6
  if d == 4:
    return 24
  if d == 5:
    return 120
  if d == 6:
    return 720
  if d == 7:
    return 5040
  if d == 8:
    return 40320
  if d == 9:
    return 362880
  raise ValueError("invalid digit")

class Problem034(Problem):

  def calculate_answer(self) -> int:
    answer = 0

    self.print_detail("Incremental Results:")

    for n in range(10, 99999):
      value = n
      digit = 0
      digit_factorial_sum = 0
      while value > 0:
        digit = value % 10
        digit_factorial_sum += digit_factorial(digit)
        value //= 10
      if n == digit_factorial_sum:
        self.print_detail(n)
        answer += n

    return answer

File: ProjectEuler/Problems/Problem034.cs
namespace ProjectEuler.Problems
{
    public class Problem034 : Problem<int>
    {
        protected override int CalculateAnswer()
        {
            var answer = 0;

            WriteLineDetail("Incremental Results:");
            for (var n = 10; n < 99999; ++n)
            {
                var value = n;
                int digit;
                var digitFactorialSum = 0;
                while (value > 0)
                {
                    digit = value % 10;
                    digitFactorialSum += DigitFactorial(digit);
                    value /= 10;
                }
                if (n == digitFactorialSum)
                {
                    WriteLineDetail(n);
                    answer += n;
                }
            }

            return answer;
        }

        private static int DigitFactorial(int digit)
        {
            if (digit == 0) return 1;
            if (digit == 1) return 1;
            if (digit == 2) return 2;
            if (digit == 3) return 6;
            if (digit == 4) return 24;
            if (digit == 5) return 120;
            if (digit == 6) return 720;
            if (digit == 7) return 5040;
            if (digit == 8) return 40320;
            if (digit == 9) return 362880;
            throw new System.Exception("invalid digit");
        }
    }
}
